ensime_shared/protocol_handler.py

```diff
@@ -0,0 +1,249 @@
+import webbrowser
+import json
+
+from ensime_shared.config import gconfig, feedback, commands
+from ensime_shared.util import catch
+from ensime_shared.symbol_format import completion_to_suggest
+
+class ProtocolHandler(object):
+
+    def __init__(self):
+        self.handlers = {}
+        self.register_responses_handlers()
+
+    def register_responses_handlers(self):
+        """Register handlers for responses from the server.
+
+        A handler must accept only one parameter: `payload`.
+        """
+        self.handlers["SymbolInfo"] = self.handle_symbol_info
+        self.handlers["IndexerReadyEvent"] = self.handle_indexer_ready
+        self.handlers["AnalyzerReadyEvent"] = self.handle_analyzer_ready
+        self.handlers["NewScalaNotesEvent"] = self.buffer_typechecks
+        self.handlers["BasicTypeInfo"] = self.show_type
+        self.handlers["ArrowTypeInfo"] = self.show_type
+        self.handlers["FullTypeCheckCompleteEvent"] = self.handle_typecheck_complete
+        self.handlers["StringResponse"] = self.handle_string_response
+        self.handlers["CompletionInfoList"] = self.handle_completion_info_list
+        self.handlers["TypeInspectInfo"] = self.handle_type_inspect
+        self.handlers["SymbolSearchResults"] = self.handle_symbol_search
+        self.handlers["DebugOutputEvent"] = self.handle_debug_output
+        self.handlers["DebugBreakEvent"] = self.handle_debug_break
+        self.handlers["DebugBacktrace"] = self.handle_debug_backtrace
+        self.handlers["DebugVmError"] = self.handle_debug_vm_error
+        self.handlers["RefactorDiffEffect"] = self.apply_refactor
+        self.handlers["ImportSuggestions"] = self.handle_import_suggestions
+        self.handlers["PackageInfo"] = self.handle_package_info
+
+    def handle_incoming_response(self, call_id, payload):
+        """Get a registered handler for a given response and execute it."""
+        self.log("handle_incoming_response: in {}".format(payload))
+        typehint = payload["typehint"]
+        handler = self.handlers.get(typehint)
+        def feature_not_supported(m):
+            msg = feedback["handler_not_implemented"]
+            self.raw_message(msg.format(typehint, self.launcher.ensime_version))
+        if handler:
+            with catch(NotImplementedError, feature_not_supported):
+                handler(call_id, payload)
```

Comment:
&lt;slight rant&gt;

> without the syntactic overhead of writing try catch all the time

I would argue that writing throwaway functions, or Python's ugly lambdas, is not significantly lighter. It makes me think more about what error handling code is trying to do, not less.

> having the error handling functions independent of the code promotes reusability and better modularity.

I cited understanding of that intent when it's appropriate, and in this particular example, the handling code *is* possibly reusable, but as written so far it's a local function so it clearly isn't reused yet. I think much more than half of the `catch` usage in the code base uses a one-time handler. If this handler above were defined at a level where it could be reused, there's no reason it couldn't be called as the one-line body of an `except` clause for the cost of one or two extra newlines.

For comparison:

```python
def feature_not_supported(m):
    msg = feedback["handler_not_implemented"]
    self.raw_message(msg.format(typehint, self.launcher.ensime_version))

if handler:
    with catch(NotImplementedError, feature_not_supported):
        handler(call_id, payload)
else:
    self.log(feedback["unhandled_response"].format(payload))
```

versus

```python
if not handler:
    self.log(feedback["unhandled_response"].format(payload))
    return

try:
    handler(call_id, payload)
except NotImplementedError:
    msg = feedback["handler_not_implemented"]
    self.raw_message(msg.format(typehint, self.launcher.ensime_version))
```

Personally I find the latter much more immediately obvious. Same number of lines, FWIW. In fairness, the former could also use an early return.

The helper function is forced to accept an argument that it doesn't need/use, making me and linters wonder if there's a bug. Also, `catch` only passes the message string of the exception object to the callable, so it's limited in a case where there is something else useful about the object. If the handler were defined non-locally, I might need to navigate away to go verify what it does. In a case like this I think that cost outweighs two lines of handler code.

`try` is a very well-understood native language feature, so it has almost zero cognitive load. `with catch` is a clever custom thing of this project that one needs to learn, and as noted above it has limitations.

Ultimately I yield to @jvican's judgment, but several instances of `catch` having been irking so I had to state my case vocally this once…